interp/numTok.ml

```diff
@@ -1,52 +1,219 @@
+
+type sign = SPlus | SMinus
+
+module SignedNat =
+struct
+  type t = sign * UnsignedNat.t
+  let of_string n =
+    let sign,n =
+      if String.length n > 0 && n.[0] = '-'
+      then (SMinus,String.sub n 1 (String.length n - 1))
+      else (SPlus,n) in
+    (sign,UnsignedNat.of_string n)
+  let to_string (s,n) =
+    (match s with SPlus -> "" | SMinus -> "-") ^ UnsignedNat.to_string n
+  let to_bigint s = Bigint.of_string (to_string s)
+  let of_bigint n =
+    let sign, n = if Bigint.is_strictly_neg n then (SMinus, Bigint.neg n) else (SPlus, n) in
+    (sign, Bigint.to_string n)
+end
+
+module Unsigned =
+struct
+
+  type t = {
+    int : string;  (** -?\[0-9\]\[0-9_\]* *)
+    frac : string;  (** empty or \[0-9_\]+ *)
+    exp : string  (** empty or \[eE\]\[+-\]?\[0-9\]\[0-9_\]* *)
+  }
+
+  let equal n1 n2 =
+    String.(equal n1.int n2.int && equal n1.frac n2.frac && equal n1.exp n2.exp)
+
+  let parse =
+    let buff = ref (Bytes.create 80) in
+    let store len x =
+      let open Bytes in
+      if len >= length !buff then
+        buff := cat !buff (create (length !buff));
+      set !buff len x;
+      succ len in
+    let get_buff len = Bytes.sub_string !buff 0 len in
+    (* reads [0-9_]* *)
+    let rec number len s = match Stream.peek s with
+      | Some ('0'..'9' as c) -> Stream.junk s; number (store len c) s
+      | Some ('_' as c) when len > 0 -> Stream.junk s; number (store len c) s
+      | _ -> len in
+    fun s ->
+    let i = get_buff (number 0 s) in
+    assert (i <> "");
+    let f =
+      match Stream.npeek 2 s with
+      | '.' :: (('0'..'9' | '_') as c) :: _ ->
+         Stream.junk s; Stream.junk s; get_buff (number (store 0 c) s)
+      | _ -> "" in
+    let e =
+      match (Stream.npeek 2 s) with
+      | (('e'|'E') as e) :: ('0'..'9' as c) :: _ ->
+         Stream.junk s; Stream.junk s; get_buff (number (store (store 0 e) c) s)
+      | (('e'|'E') as e) :: (('+'|'-') as sign) :: _ ->
+         begin match Stream.npeek 3 s with
+         | _ :: _ :: ('0'..'9' as c) :: _ ->
+            Stream.junk s; Stream.junk s; Stream.junk s;
+            get_buff (number (store (store (store 0 e) sign) c) s)
+         | _ -> ""
+         end
+      | _ -> "" in
+    { int = i; frac = f; exp = e }
+
+  let to_string n =
+    n.int ^ (if n.frac = "" then "" else "." ^ n.frac) ^ n.exp
+
+  let of_string s =
+    if s = "" || s.[0] < '0' || s.[0] > '9' then None else
+      let strm = Stream.of_string (s ^ "  ") in
+      let n = parse strm in
+      if Stream.count strm >= String.length s then Some n else None
+
+  let to_nat ?loc = function
+    | { int = i; frac = ""; exp = "" } -> i
+    | _ -> CErrors.user_err ?loc (Pp.str "This number is not a natural number.")
+
+  let is_nat = function
+    | { int = _; frac = ""; exp = "" } -> true
+    | _ -> false
+
+end
+
+open Unsigned
+
+module Signed =
+struct
+
+  type t = sign * Unsigned.t
+
+  let equal (s1,n1) (s2,n2) =
+    s1 = s2 && equal n1 n2
+
+  let is_zero s =
```

Comment:
This can probably be implemented using `UnsignedNat.compare` above, saving a few lines.

Reply:
Good idea, I did it. (Please check though.)

Reply:
Which compare is this? (I don't see any compare in Unsigned)

Reply:
Indeed. This was the wrong `compare` (from `Pervasives`) and, more than that, I completely broke the code. Hopefully fixed.